source/include/microapp/cs_MicroappStorage.h

```diff
@@ -2,148 +2,143 @@
-		 */
-		uint32_t handlePacket(microapp_packet_header_t *packet_stub);
-
-		/**
-		 * When fstorage is done, this function will be called (indirectly through app_scheduler).
-		 */
-		void handleFileStorageEvent(nrf_fstorage_evt_t *evt);
-
-		/**
-		 * Get header of microapp.
-		 */
-		void getHeaderApp(microapp_header_t *header);
-
-		/**
-		 * Check if app is valid (only checking the FDS record field).
-		 */
-		bool isAppValid();
-
-		/**
-		 * Validate the overall binary. This goes through flash and checks it completely. All flash write operations
-		 * have to have finished before.
-		 */
-		uint16_t validateApp();
-
-		/**
-		 * Set app to be valid (only the FDS record field).
-		 */
-		void setAppValid();
-
-		/**
-		 * Enable app.
-		 */
-		uint16_t enableApp(bool flag);
-
-		/**
-		 * Check if app is enabled.
-		 */
-		bool isEnabled();
-
-		/**
-		 * Check if app will fit in the reserved pages.
-		 */
-		uint16_t checkAppSize(uint16_t size);
-
-		/**
-		 * Write a chunk to flash memory. Writes to buffer at index start + index * CHUNK_SIZE up to size of the data.
-		 */
-		uint16_t writeChunk(uint8_t index, const uint8_t *data, uint16_t size);
-
-		/**
-		 * Validate chunk by comparing its calculated checksum with the checksum in the packet.
-		 */
-		uint16_t validateChunk(const uint8_t * const data, uint16_t size, uint16_t compare);
-
-		/**
-		 * Store in flash information about the app (start address, checksum, etc.)
-		 */
-		void storeAppMetadata(uint8_t id, uint16_t checksum, uint16_t size);
-
-
+class MicroappStorage {
+public:
+	static MicroappStorage& getInstance() {
+		static MicroappStorage instance;
+		return instance;
+	}
+
+	/**
+	 * Initialize fstorage, allocate buffer.
+	 */
+	cs_ret_code_t init();
+
+	/**
+	 * Get a copy of the microapp binary header.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
+	 * @param[out] header    Pointer where will be read to.
+	 */
+	void getAppHeader(uint8_t appIndex, microapp_binary_header_t* header);
+
+	/**
+	 * Get the address in flash, where the microapp program starts.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
+	 */
+	uint32_t getStartInstructionAddress(uint8_t appIndex);
+
+	/**
+	 * Checks if storage space of this microapp is erased.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
+	 * @return true                         Storage space of this app is erased, and thus ready to be written.
+	 */
+	bool isErased(uint8_t appIndex);
+
+	/**
+	 * Erases storage space of given app.
+	 * First checks if storage is already erased.
+	 *
+	 * @param[in] appIndex   Index of the microapp, validity is not checked.
+	 *
+	 * @return ERR_SUCCESS_NO_CHANGE        The storage is already erased.
+	 * @return ERR_WAIT_FOR_SUCCESS         The storage will be written erased, wait for EVT_MICROAPP_ERASE_RESULT.
+	 */
+	cs_ret_code_t erase(uint8_t appIndex);
```

Comment:
undefined behaviour when appIndex is not valid? Can we say anything about this before we forget the implementation details and it really is impossible to find out unless rereading the code?

Reply:
It will probably lead to a simple nrf error, because you'll try to read outside the space that's claimed by this fstorage instance.
The whole point of mentioning that appIndex is not checked, is that the caller should do that.